aptos-move/aptos-vm-types/src/resource_group_adapter.rs

```diff
@@ -0,0 +1,525 @@
+    ) -> Self {
+        Self {
+            resource_resolver: UnifiedResourceView::ResourceView(resource_view),
+            group_size_kind,
+            group_cache: RefCell::new(HashMap::new()),
+        }
+    }
+
+    pub fn from_state_view(
+        state_view: &'r dyn TStateView<Key = K>,
+        group_size_kind: GroupSizeKind,
+    ) -> Self {
+        Self {
+            resource_resolver: UnifiedResourceView::StateView(state_view),
+            group_size_kind,
+            group_cache: RefCell::new(HashMap::new()),
+        }
+    }
+
+    pub fn group_size_kind(&self) -> GroupSizeKind {
+        self.group_size_kind.clone()
+    }
+
+    fn release_group_cache(&self) -> HashMap<K, BTreeMap<T, Bytes>> {
+        self.group_cache
+            .borrow_mut()
+            .drain()
+            .map(|(k, v)| (k, v.0))
+            .collect()
+    }
+
+    // Ensures that the resource group at state_key is cached in self.group_cache. Ok(true)
+    // means the resource was already cached, while Ok(false) means it just got cached.
+    fn ensure_cached(&self, state_key: &K) -> anyhow::Result<bool> {
+        if self.group_cache.borrow().contains_key(state_key) {
+            return Ok(true);
+        }
+
+        let group_data_from_resolver = self.resource_resolver.get_bytes(state_key)?;
+        let (group_data, blob_len): (BTreeMap<T, Bytes>, u64) = group_data_from_resolver
+            .map_or_else(
+                || Ok::<_, Error>((BTreeMap::new(), 0)),
+                |group_data_blob| {
+                    let group_data = bcs::from_bytes(&group_data_blob)
+                        .map_err(|_| anyhow::Error::msg("Resource group deserialization error"))?;
+                    Ok((group_data, group_data_blob.len() as u64))
+                },
+            )?;
+
+        let group_size = match self.group_size_kind {
+            GroupSizeKind::None => 0,
+            GroupSizeKind::AsBlob => blob_len,
+            GroupSizeKind::AsSum => group_data
+                .iter()
+                .try_fold(0, |len, (tag, res)| {
+                    let delta = bcs::serialized_size(tag)? + res.len();
+                    Ok(len + delta as u64)
+                })
+                .map_err(|_: Error| {
+                    anyhow::Error::msg("Resource group member tag serialization error")
+                })?,
+        };
+
+        self.group_cache
+            .borrow_mut()
+            .insert(state_key.clone(), (group_data, group_size));
+        Ok(false)
+    }
+}
+
+pub type ResourceGroupAdapter<'a> = TResourceGroupAdapter<'a, StateKey, StructTag, MoveTypeLayout>;
+
+impl<K, T, L> TResourceGroupView for TResourceGroupAdapter<'_, K, T, L>
+where
+    K: Clone + Eq + Hash,
+    T: Debug + DeserializeOwned + Ord + Serialize,
+{
+    type Key = K;
+    type Layout = L;
+    type Tag = T;
+
+    fn resource_group_size(&self, state_key: &Self::Key) -> anyhow::Result<u64> {
+        self.ensure_cached(state_key)?;
+        Ok(self
```

Comment:
possible to return refs to the value from `ensure_cached()`?

Reply:
I tried, it complains about returning from data owned by another function, the data structure lives in RefCell and I couldn't make it work?

Reply:
if it can be done and you or @zekun000 can show me how to do it, I'll be grateful ^